src/include/platform/internal/GenericPlatformManagerImpl_Zephyr.ipp

```diff
@@ -194,10 +226,74 @@ void GenericPlatformManagerImpl_Zephyr<ImplClass>::_RunEventLoop(void)
         ProcessDeviceEvents();
     }
     SystemLayerSocketsLoop().EventLoopEnds();
+}
 
-    Impl()->UnlockChipStack();
+template <class ImplClass>
+void GenericPlatformManagerImpl_Zephyr<ImplClass>::ProcessDeviceEvents()
+{
+    ChipDeviceEvent event;
+
+    while (k_msgq_get(&mChipEventQueue, &event, K_NO_WAIT) == 0)
+    {
+        SYSTEM_STATS_DECREMENT(System::Stats::kPlatformMgr_NumEvents);
+        Impl()->DispatchEvent(&event);
+    }
+}
+
+#else
+
+template <class ImplClass>
+void GenericPlatformManagerImpl_Zephyr<ImplClass>::ProcessEventLoop()
+{
+    System::LayerImplZephyr & systemLayer = static_cast<System::LayerImplZephyr &>(DeviceLayer::SystemLayer());
+
+    while (mShouldRunEventLoop)
+    {
+        k_timeout_t timeout = K_FOREVER;
+        bool eventReceived  = false;
+        ChipDeviceEvent event;
+
+        // Get the awaken time for the next timer
+        std::optional<System::Clock::Timestamp> nextAwakenTime = systemLayer.GetNextAwakenTime();
+        if (nextAwakenTime.has_value())
+        {
+            System::Clock::Timestamp now = System::SystemClock().GetMonotonicTimestamp();
+            if (nextAwakenTime.value() > now)
+            {
+                // Calculate the timeout
+                System::Clock::Timestamp delay = nextAwakenTime.value() - now;
+                timeout                        = K_MSEC(delay.count());
+            }
+            else
+            {
+                // Some timer already expired, do not wait if event queue is empty
+                timeout = K_NO_WAIT;
+            }
+        }
+
+        Impl()->UnlockChipStack();
+        eventReceived = k_msgq_get(&mChipEventQueue, &event, timeout) == 0;
+        Impl()->LockChipStack();
+
+        // Call into the system layer to dispatch the callback functions for all timers
+        // that have expired.
+        CHIP_ERROR err = systemLayer.HandlePlatformTimer();
```

Comment:
Note that you call this every time `k_msgq_get` exits, even if the function has exited because of an event, not timeout. This makes some calls to `HandlePlatformTimer` unnecessary. This could be avoided by moving this to line 269 above.